platforms/chibios/drivers/vendor/RP/RP2040/ps2_vendor.c

```diff
@@ -0,0 +1,335 @@
+// Copyright 2022 Marek Kraus (@gamelaster)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "quantum.h"
+
+#include "hardware/pio.h"
+#include "hardware/clocks.h"
+#include "ps2.h"
+#include "print.h"
+
+#if !defined(MCU_RP)
+#    error PIO Driver is only available for Raspberry Pi 2040 MCUs!
+#endif
+
+#if defined(PS2_ENABLE)
+#    if defined(PS2_MOUSE_ENABLE)
+#        if !defined(PS2_MOUSE_USE_REMOTE_MODE)
+#            define BUFFERED_MODE_ENABLE
+#        endif
+#    else // PS2 Keyboard
+#        define BUFFERED_MODE_ENABLE
+#    endif
+#endif
+
+#if PS2_DATA_PIN + 1 != PS2_CLOCK_PIN
+#    error PS/2 Clock pin must be followed by data pin!
+#endif
+
+static inline void pio_serve_interrupt(void);
+
+#if defined(PS2_PIO_USE_PIO1)
+static const PIO pio = pio1;
+
+OSAL_IRQ_HANDLER(RP_PIO1_IRQ_0_HANDLER) {
+    OSAL_IRQ_PROLOGUE();
+    pio_serve_interrupt();
+    OSAL_IRQ_EPILOGUE();
+}
+#else
+static const PIO pio = pio0;
+
+OSAL_IRQ_HANDLER(RP_PIO0_IRQ_0_HANDLER) {
+    OSAL_IRQ_PROLOGUE();
+    pio_serve_interrupt();
+    OSAL_IRQ_EPILOGUE();
+}
+#endif
+
+#define PS2_WRAP_TARGET 2
+#define PS2_WRAP 30
+
+// clang-format off
+static const uint16_t ps2_program_instructions[] = {
+    0xe080, //  0: set    pindirs, 0
+    0xe000, //  1: set    pins, 0
+            //     .wrap_target
+    0xe020, //  2: set    x, 0
+    0x8080, //  3: pull   noblock
+    0xa027, //  4: mov    x, osr
+    0x0048, //  5: jmp    x--, 8
+    0x00c2, //  6: jmp    pin, 2
+    0x0017, //  7: jmp    23
+    0xe082, //  8: set    pindirs, 2
+    0xef00, //  9: set    pins, 0                [15]
+    0xaf42, // 10: nop                           [15]
+    0xe283, // 11: set    pindirs, 3             [2]
+    0xe081, // 12: set    pindirs, 1
+    0xe002, // 13: set    pins, 2
+    0xe029, // 14: set    x, 9
+    0x2021, // 15: wait   0 pin, 1
+    0x6601, // 16: out    pins, 1                [6]
+    0x004f, // 17: jmp    x--, 15
+    0xe080, // 18: set    pindirs, 0
+    0xe001, // 19: set    pins, 1
+    0x2021, // 20: wait   0 pin, 1
+    0x20a1, // 21: wait   1 pin, 1
+    0x0002, // 22: jmp    2
+    0xe080, // 23: set    pindirs, 0
+    0x4001, // 24: in     pins, 1
+    0x20a1, // 25: wait   1 pin, 1
+    0xe029, // 26: set    x, 9
+    0x2021, // 27: wait   0 pin, 1
+    0x4001, // 28: in     pins, 1
+    0x20a1, // 29: wait   1 pin, 1
+    0x005b, // 30: jmp    x--, 27
+            //     .wrap
+};
+// clang-format on
+
+static const struct pio_program ps2_program = {
+    .instructions = ps2_program_instructions,
+    .length       = 31,
+    .origin       = -1,
+};
+
+static int         state_machine = -1;
+thread_reference_t rx_thread     = NULL;
+thread_reference_t tx_thread     = NULL;
+
+#ifdef BUFFERED_MODE_ENABLE
+#    define PBUF_SIZE 32
+static uint32_t pbuf[PBUF_SIZE];
+static uint8_t  pbuf_head = 0;
+static uint8_t  pbuf_tail = 0;
+
+static inline void     pbuf_enqueue(uint32_t frame);
+static inline uint32_t pbuf_dequeue(void);
```

Comment:
No need to implement your own buffer here you can use [ChibiOS I/O bytes queues](http://chibiforge.org/doc/21.11/full_rm/group___h_a_l___q_u_e_u_e_s.html)

Reply:
Oh, good to know! Will rework it to use them instead :blush:

Reply:
Sorry I meant an input queue, as we want to fill it from the ISR and read it from the main thread.